### ISSUE TITLE
Wazuh vulnerability detector shows only wazuh-manager-master (agent 000)

### ISSUE THREAD

**Issue by lluczaj** [wazuh/wazuh#11136]
Wazuh installed on Kubernetes (wazuh/wazuh-odfe:4.2.5, amazon/opendistro-for-elasticsearch:1.13.2)
Agents installed from rpm (4.2.5) 

Wazuh agents can register in wazuh cluster. Inventory is available and all packages are listed. However, Wazuh vulnerability detector shows only Agent 000 (wazuh-manager-master). After enabling debug logs I can see the following for each agent except 000 in ossec.log:

```
2021/11/30 15:57:35 wazuh-modulesd:database[703] wm_database.c:276 at wm_sync_agents(): DEBUG: Synchronizing agent 171 'ip-xxxxxxx.region.compute.internal'.
2021/11/30 15:57:35 wazuh-modulesd[703] wdb_global_helpers.c:103 at wdb_insert_agent(): DEBUG: Global DB Error reported in the result of the query
2021/11/30 15:57:35 wazuh-modulesd:database[703] wm_database.c:470 at wm_sync_agent_group(): DEBUG: wm_sync_agent_group(171): 0.000 ms.
2021/11/30 15:57:35 wazuh-modulesd:database[703] wm_database.c:276 at wm_sync_agents(): DEBUG: Synchronizing agent 172 'ip-xxxxxxx.region.compute.internal'.
2021/11/30 15:57:35 wazuh-modulesd[703] wdb_global_helpers.c:103 at wdb_insert_agent(): DEBUG: Global DB Error reported in the result of the query
2021/11/30 15:57:35 wazuh-modulesd:database[703] wm_database.c:470 at wm_sync_agent_group(): DEBUG: wm_sync_agent_group(172): 0.000 ms.
[...]
```

Vulnerability detector is enabled on master:
```
  <vulnerability-detector>
      <enabled>yes</enabled>
      <interval>60m</interval>
      <ignore_time>6h</ignore_time>
      <run_on_start>yes</run_on_start>
```

Syscollector is enabled on agents:
```
 <wodle name="syscollector">
    <disabled>no</disabled>
    <interval>1h</interval>
    <scan_on_start>yes</scan_on_start>
    <hardware>yes</hardware>
    <os>yes</os>
    <network>yes</network>
    <packages>yes</packages>
    <ports all="no">yes</ports>
    <processes>yes</processes>
  </wodle>
```

Can anyone help to track down this issue ?

**Comment by MiguelazoDS** [comment]
Hello @lluczaj, 

I was able to replicate those error messages locally with ease, I suspect those are harmless, but in that case, they shouldn't be reported as "Error", which is weird. 
To be completely sure could you put the wazuh_db debug in 1. You already did with wazuh_modules, but just in case you should add the line `wazuh_db.debug=1` in the `/var/ossec/etc/local_internal_options` (or internal_options, if you don't mind the config getting erased on upgrades). 

The line I'm looking for is something like this. 

`2021/11/30 17:03:53 wazuh-db[92082] wdb_parser.c:4640 at wdb_parse_global_insert_agent(): DEBUG: Global DB Cannot execute SQL query; err database queue/db/global.db: UNIQUE constraint failed: agent.id`

It should be just above the error messages you shared.

To get those logs I restart the manager and during the start process, there's a check that looks for agents. It attempts to insert them but since the ID is the same it rejects the registration. And again, I didn't notice this behavior before so I probably need to check it with my team. 

On the other hand, besides those error messages, I have the agents connected and active, that's why I'm thinking that we should change those messages or improve that behavior. 

You can check that with `agent_control -l` on your manager. 

![image](https://user-images.githubusercontent.com/13010397/144120422-a3d888dd-a13a-4f12-a32b-b2bb1ba23a5e.png)

I commented that because the vulnerability detector checks for active agents before starting any scan. 

Regards!

**Comment by lluczaj** [comment]
Hi @MiguelazoDS, thanks for your reply! I enabled debug logs as you asked and that's what I get:

```
2021/12/01 07:32:37 wazuh-modulesd:database[679] wm_database.c:470 at wm_sync_agent_group(): DEBUG: wm_sync_agent_group(110): 0.000 ms.
2021/12/01 07:32:37 wazuh-modulesd:database[679] wm_database.c:276 at wm_sync_agents(): DEBUG: Synchronizing agent 111 'ip-xxxxxxx.eu-west-1.compute.internal'.
2021/12/01 07:32:37 wazuh-db[466] wdb.c:1165 at wdb_stmt_cache(): DEBUG: DB(global) sqlite3_reset() stmt(95): UNIQUE constraint failed: agent.id
2021/12/01 07:32:37 wazuh-db[466] wdb_global.c:89 at wdb_global_insert_agent(): DEBUG: SQLite: UNIQUE constraint failed: agent.id
2021/12/01 07:32:37 wazuh-db[466] wdb_parser.c:4784 at wdb_parse_global_insert_agent(): DEBUG: Global DB Cannot execute SQL query; err database queue/db/global.db: UNIQUE constraint failed: agent.id
2021/12/01 07:32:37 wazuh-modulesd[679] wdb_global_helpers.c:103 at wdb_insert_agent(): DEBUG: Global DB Error reported in the result of the query
```

The line you were looking for indeed appears in logs. 

`agent-control -l`
```
Wazuh agent_control. List of available agents:
   ID: 000, Name: wazuh-manager-master-0 (server), IP: 127.0.0.1, Active/Local
   ID: 001, Name: ip-xxxx.eu-west-1.compute.internal, IP: any, Disconnected
   ID: 002, Name: xxxx.internal.nth.local, IP: any, Active
   ID: 003, Name: ip-xxxxx.eu-west-1.compute.internal, IP: any, Active
   ID: 004, Name: xxxx.internal.nth.local, IP: any, Active
   ID: 005, Name: xxxx.prod.nth.local, IP: any, Active
   [...]
```

All connected agents are visible in Kibana, I can access their inventory and see a package list. All other modules seems to work correctly (like integrity checks). However there is nothing for them in vulnerability scanning section, there's just `agent 000`. Also it doesn't seem that they are analyzed at all based on logs, there is only something for master but no other agents:

```
2021/12/01 07:34:35 wazuh-modulesd:vulnerability-detector[679] wm_vuln_detector.c:2167 at wm_vuldet_check_agent_vulnerabilities(): INFO: (5450): Analyzing agent '000' vulnerabilities.
2021/12/01 07:34:35 wazuh-modulesd:vulnerability-detector[679] wm_vuln_detector.c:4446 at wm_vuldet_get_software_info(): DEBUG: (5437): Collecting agent '000' software.
2021/12/01 07:34:35 wazuh-modulesd:vulnerability-detector[679] wm_vuln_detector.c:4464 at wm_vuldet_get_software_info(): DEBUG: (5439): A partial scan will be run on agent '000'
```

**Comment by MiguelazoDS** [comment]
@lluczaj, sorry for the late response. What OS is installed on those agents? I'm thinking that probably there's no data for those agents. 

We should see the "Analyzing agent" log, but then I'm looking for a log like this. 

`2021/12/01 11:46:47 wazuh-modulesd:vulnerability-detector[1817] wm_vuln_detector.c:2480 at wm_vuldet_check_agent_vulnerabilities(): WARNING: (5575): Unavailable vulnerability data for the agent '000' OS. Skipping it.`

(`grep -i "unavailable vulnerability" /var/ossec/logs/ossec.log`)

In this case, I did not enable the provider for centos (which is my manager), so the `/var/ossec/queue/vulnerabilities/cve.db` is empty for that OS.

That is configurable in `/var/ossec/etc/ossec.conf` in the **provider** section inside **vulnerability-detector** tag as explained here. 

https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/vuln-detector.html#provider

As an example, if your agent is running on a Debian machine you should configure this section. 

```xml
<!-- Debian OS vulnerabilities -->
<provider name="debian">
  <enabled>no</enabled>
  <os>stretch</os>
  <os>buster</os>
  <update_interval>1h</update_interval>
</provider>
```

Another thing that could be happening is that you are using an unsupported OS, if that is the case you could check this section. 

https://documentation.wazuh.com/current/user-manual/capabilities/vulnerability-detection/allow-os.html

To verify if that is the case use something like this to find the log. 

`grep -i "unsupported OS" /var/ossec/logs/ossec.log`

**Comment by MiguelazoDS** [comment]
@lluczaj 

Yes, you're right about this `I thought that both CentOS and Amazon Linux could somehow fall under RedHat provider.` There're some changes targeted to the next release that includes native support for Amazon Linux, you can check the details here if you want https://github.com/wazuh/wazuh/issues/6734, but besides that, you should be able to scan those agents as it is now.

Vulnerability-detector first looks for connected agents and then check if there's CVE information for them, so I would like to see the complete logs related to vulnerability-detector. If those agents are not being scanned some information should be present in the log file. 
Could you share the /var/ossec/log/ossec.log file? 
It would be nice to keep the wazuh_modules.debug=2 and perform a manage restart so if the `<run_on_start>yes</run_on_start>` option is configured it should trigger a scan. After that, I'm interested in everything between these two lines. 

`2021/12/02 14:11:48 wazuh-modulesd:vulnerability-detector[20007] wm_vuln_detector.c:7640 at wm_vuldet_run_scan(): INFO: (5431): Starting vulnerability scan.`

`2021/12/02 14:11:48 wazuh-modulesd:vulnerability-detector[20007] wm_vuln_detector.c:7655 at wm_vuldet_run_scan(): INFO: (5472): Vulnerability scan finished.`

Also if you could run this query. 

`sqlite3 /var/ossec/queue/vulnerabilities/cve.db 'select target,count(*) from vulnerabilities group by target'` 

To be sure the CVE information is available. 

I hope we can find the root cause of this. Regards!

**Comment by lluczaj** [comment]
@MiguelazoDS 

Below is the log that you asked:
```
2021/12/03 14:09:58 wazuh-modulesd:vulnerability-detector[665] wm_vuln_detector.c:2167 at wm_vuldet_check_agent_vulnerabilities(): INFO: (5450): Analyzing agent '000' vulnerabilities.
2021/12/03 14:09:58 wazuh-modulesd:vulnerability-detector[665] wm_vuln_detector.c:4446 at wm_vuldet_get_software_info(): DEBUG: (5437): Collecting agent '000' software.
2021/12/03 14:09:58 wazuh-modulesd:vulnerability-detector[665] wm_vuln_detector.c:4464 at wm_vuldet_get_software_info(): DEBUG: (5439): A partial scan will be run on agent '000'
2021/12/03 14:09:58 wazuh-modulesd:vulnerability-detector[665] wm_vuln_detector.c:4672 at wm_vuldet_get_software_info(): DEBUG: (5445): No changes have been found with respect to the last package inventory or no packages have been indexed for agent '000'
2021/12/03 14:09:58 wazuh-modulesd:vulnerability-detector[665] wm_vuln_detector.c:2216 at wm_vuldet_check_agent_vulnerabilities(): INFO: (5471): Finished vulnerability assessment for agent '000'
2021/12/03 14:09:58 wazuh-modulesd:vulnerability-detector[665] wm_vuln_detector.c:2217 at wm_vuldet_check_agent_vulnerabilities(): DEBUG: (5470): It took '0' seconds to 'scan' vulnerabilities in agent '000'
2021/12/03 14:09:58 wazuh-modulesd:vulnerability-detector[665] wm_vuln_detector.c:7144 at wm_vuldet_run_scan(): INFO: (5472): Vulnerability scan finished.
```

Query:
```
[wazuh-manager-master-0 logs]# sqlite3 /var/ossec/queue/vulnerabilities/cve.db 'select target,count(*) from vulnerabilities group by target'
BIONIC|71815
BUSTER|24590
FOCAL|39787
RHEL5|24936
RHEL6|76984
RHEL7|84833
RHEL8|75183
STRETCH|27405
TRUSTY|55572
XENIAL|62696
```

**Comment by MiguelazoDS** [comment]
Hi @lluczaj, 

It seems that the solution is not as trivial as I expected due to some lack of configuration for unsupported OS, feeds information, or similar. 
What I can see from those logs is that the task in charge to collects the agent, prior to the scan, is not taking into account the rest of the agents.

The module first looks for connected agents and then collects the information for each of them. Some of those steps are failing, and according to the information you shared before, it is probably the second one. 

I bother you again to share the logs, but this time for the complete ossec.log with wazuh_modules.debug=2 and wazuh_db.debug=2. Please, if you have some sensitive information erase it before sharing your logs.
Also, I would like to take a look at the `agent` table in your global database.

Please send me the output of 

sqlite3 /var/ossec/queue/db/global.db 'select * from agent'

Sorry for the inconvenience.

**Comment by lluczaj** [comment]
Hi @MiguelazoDS I would prefer not to share the full log here. It's huge and its containing plenty of info about our infrastructure so even after I think I erased all the sensitive information from it I cannot be 100% sure. Are there any specific lines that you're looking for or maybe we can narrow it to a single agent somehow ?

Partial output of sqlite command (we have 44 connected agents). Also, I omitted disconnected agents as we have plenty of them due to some Kubernetes clusters on spot instances. 

```
[wazuh-manager-master-0 /]# sqlite3 /var/ossec/queue/db/global.db 'select * from agent' | grep -v disconne
0|wazuh-manager-master-0|127.0.0.1|127.0.0.1||CentOS Linux|7.9|7|9|||centos|Linux |wazuh-manager-master-0 |5.4.156-83.273.amzn2.x86_64 |#1 SMP Sat Oct 30 12:59:07 UTC 2021 |x86_64|x86_64|Wazuh v4.2.5|||wazuh-manager-master-0|wazuh-manager-master|1636966146|253402300799||synced|active
2|xyz.internal.xyz.local|xx.yy.133.78|any|0db4a14fcf7c6c43024ad1fdd03c98cdd8e1a6a04a232645df0b3a7b5d4b0b3f|CentOS Linux|7.9|7|9|||centos|Linux |xyz.internal.xyz.local |3.10.0-1160.45.1.el7.x86_64 |#1 SMP Wed Oct 13 17:20:51 UTC 2021 |x86_64|x86_64|Wazuh v4.2.5|ab73af41699f13fdd81903b5f23d8d00|2c45c95db2954d2c7d0ea533f09e81a5|wazuh-manager-worker-0|wazuh-manager-worker-0|1636966149|1638778526|default|synced|active
3|ip-xx-yy-133-19.region.compute.internal|xx.yy.133.19|any|8ed8e0dfe7400e4b3fdb1861994472fb5d7246cc5f649689d38c2cd97499f582|CentOS Linux|7.9|7|9|||centos|Linux |ip-xx-yy-133-19.region.compute.internal |3.10.0-1160.45.1.el7.x86_64 |#1 SMP Wed Oct 13 17:20:51 UTC 2021 |x86_64|x86_64|Wazuh v4.2.5|ab73af41699f13fdd81903b5f23d8d00|2c45c95db2954d2c7d0ea533f09e81a5|wazuh-manager-worker-0|wazuh-manager-worker-0|1636966149|1638778524|default|synced|active
4|xyz.internal.xyz.local|xx.yy.133.212|any|76da3fa3b681d7a7dfa0eb74e0d541b8b16849b58037624283d2cd196435c311|CentOS Linux|7.9|7|9|||centos|Linux |xyz.internal.xyz.local |3.10.0-1160.45.1.el7.x86_64 |#1 SMP Wed Oct 13 17:20:51 UTC 2021 |x86_64|x86_64|Wazuh v4.2.5|ab73af41699f13fdd81903b5f23d8d00|2c45c95db2954d2c7d0ea533f09e81a5|wazuh-manager-worker-0|wazuh-manager-worker-0|1636966149|1638778529|default|synced|active
5|xyz.prod.xyz.local|xx.yy.131.147|any|3111805f21e74d5e54f650368e1587d198171b6d27ef51b66fe66d5688bc40f8|CentOS Linux|7.9|7|9|||centos|Linux |xyz.prod.xyz.local |3.10.0-1160.45.1.el7.x86_64 |#1 SMP Wed Oct 13 17:20:51 UTC 2021 |x86_64|x86_64|Wazuh v4.2.5|ab73af41699f13fdd81903b5f23d8d00|2c45c95db2954d2c7d0ea533f09e81a5|wazuh-manager-worker-1|wazuh-manager-worker-1|1636966149|1638778518|default|synced|active
```

**Comment by MiguelazoDS** [comment]
Hi @lluczaj,

I understand it, and in fact, the logs I was expecting must have been in the logs you shared before, it was just to do an overall lookup. It's strange that we are not getting any log of why the agents are not being detected. I will try to replicate a local environment with that information. 
To simplify that task, could you share the `sys_osinfo` information from those agents? the vulnerability-detector module gets information from the agent table in the global database, as well as the `sys_osinfo` information from the agent database to perform the scans. What I see is that the worker hostname is the same as the node name, but I need to confirm if that is really the problem here.

The query is the following. Since all of them have the same problem we will try with any. 

`sqlite3 /var/ossec/queue/db/002.db 'select * from sys_osinfo'`

What OS are your workers? Are they also using Wazuh 4.2.5? 

Thanks for the information provided.

**Comment by lluczaj** [comment]
@MiguelazoDS 
```
sqlite3 /var/ossec/queue/db/002.db 'select * from sys_osinfo'
Error: no such table: sys_osinfo
```

002.db seems to be empty. There are only some data in 000.db.
```
du -hs /var/ossec/queue/db/*

976K    /var/ossec/queue/db/000.db
0       /var/ossec/queue/db/002.db
256K    /var/ossec/queue/db/global.db
132K    /var/ossec/queue/db/global.db-journal
0       /var/ossec/queue/db/wdb

sqlite3 /var/ossec/queue/db/000.db 'select * from sys_osinfo'
0|2021/12/07 13:42:46|wazuh-manager-master-0|x86_64|Centos Linux|7.9.2009|Core|7|9|2009||centos|Linux|5.4.156-83.273.amzn2.x86_64|#1 SMP Sat Oct 30 12:59:07 UTC 2021||1638884565427101318
```

Also I have just added Ubuntu worker and it seems that it also isn't scanned for vulnerabilities. There are multiple OS and most of them use wazuh agent 4.2.5. 

OS: Centos Linux 7.9.2009
OS: Amazon Linux 2
OS: Ubuntu 18.04.4 LTS (Bionic Beaver)

**Comment by MiguelazoDS** [comment]
Hi @lluczaj,

I was overlooking a completely basic thing that I was not aware of, and the team make me notice that in a cluster environment the vulnerability-detector module should be configured on the worker node. So that manager will scan every agent directly connected to it. 
In your case, there's no agent connected directly to your master node. 

`002.db seems to be empty. There is only some data in 000.db.` exactly, I deleted my question hoping you were not wasting time getting that information. Because those tables are on your worker nodes. That's why we could not find any log related to agents, because the master has not any agent connected directly. 

Unfortunately, there's no such thing as **centralized configuration** for workers, so you must achieve the configuration manually on every worker. Please perform the vulnerability-detector configuration on your workers and let us know how it goes. And if you experience more issues the logs we must see are the ones from your worker.

**Comment by lluczaj** [comment]
Thanks @MiguelazoDS I will test it today and get back to you with results !

**Comment by MiguelazoDS** [comment]
I hope we can find the issue. Looking forward to your news.

**Comment by MiguelazoDS** [comment]
Closing issue due to inactivity. Don't hesitate to re-open if needed.

**Comment by redsigma** [comment]
I have a similar problem with vulnerabiliy-detector only analyzing agent '000` which is wazuh-manager-master 

I have configured the `ossec.conf` file ( on wazuh-manager-master side)  with the proper `<os allow=>` settings. I have checked them with the console api and they are correct.

I tried to add the same configuration of vulnerabiltiy-detector in the `ossec.conf` file on the agent side, and when i restart the agent it logs this:

config.c:164 at read_main_elements(): WARNING: vulnerability-detector configuration is only set in the manager.

Does this mean that vulnerability-detector works only with the wazuh-manager-master ?
